### PR TITLE
[13.0][IMP] dms: Inherit permissions from related records

### PR DESCRIPTION
**Pull request by victoralmau**
FWP from 12.0: https://github.com/OCA/dms/pull/68

Add Inherit access from parent record (in storage).

Locked by:
- [x] https://github.com/OCA/dms/pull/67

Please @pedrobaeza and @Yajo can you review it?

@Tecnativa TT29041

### CODE REVIEW

**Comment by pedrobaeza** [comment]
I think this should wait until #67 is merged and the code is simplified.

**Comment by victoralmau** [comment]
It's ready to merge now, right?

File: dms/controllers/portal.py [hunk 1]
```diff
@@ -16,24 +16,22 @@
...
         except (AccessError, MissingError):
             return False
 
         if access_token:
-            if not item_sudo.check_access_token(access_token):
+            if not item.check_access_token(access_token):
                 return False
         else:
-            if not item_sudo.with_user(request.env.user.id).check_access("read", False):
+            if not item.sudo(request.env.user.id).check_access("read", False):
```
Comment (yajo) [inline dms/controllers/portal.py]: This should be `with_user`, isn't it?

File: dms/models/dms_file.py [hunk 2]
```diff
@@ -131,6 +131,19 @@ class File(models.Model):
         attachment=True, string="Content File", prefetch=False, invisible=True
     )
 
+    def check_access_rule(self, operation):
+        self.mapped("directory_id").check_access_rule(operation)
+        return super().check_access_rule(operation)
+
+    def get_human_size(self):
+        return human_size(self.size)
```
Comment (yajo) [inline dms/models/dms_file.py]: It seems this diff is unnecessary if you keep the method where it was.

File: dms/models/storage.py [hunk 3]
```diff
@@ -85,6 +85,19 @@ class Storage(models.Model):
     )
 
     model_ids = fields.Many2many("ir.model", string="Linked Models")
+    inherit_access_from_parent_record = fields.Boolean(
+        string="Inherit permissions from related record",
+        default=False,
```
Comment (yajo) [inline dms/models/storage.py:91]: IMHO this is the most sane default (see TT29500).

```suggestion
        default=True,
```

File: dms/models/storage.py [hunk 4]
```diff
@@ -85,6 +85,19 @@ class Storage(models.Model):
...
+        help="Indicate if directories and files access work only with "
+        "related model access (for example, if some directories are related "
+        "with any sale, only users with read access to these sale can acess)",
+    )
+    include_message_attachments = fields.Boolean(
+        string="Create files from message attachments",
+        default=False,
```
Comment (yajo) [inline dms/models/storage.py:98]: Same here:

```suggestion
        default=True,
```

**Review by yajo — CHANGES_REQUESTED** [review]
Some little details, but it looks good overall to me.

I'd appreciate to see more coverage, given the importance of this feature.

File: dms/models/storage.py (hunk 3 above)
Comment (pedrobaeza) [reply dms/models/storage.py:91]: No, this shouldn't be enabled by default, as it's an opt-in feature, and even more, you will be changing current behavior on existing installations.

File: dms/models/storage.py (hunk 4 above)
Comment (pedrobaeza) [reply dms/models/storage.py:98]: The same as above.

File: dms/models/storage.py (hunk 3 above)
Comment (yajo) [reply dms/models/storage.py:91]: Setting `False` here means that you want to bypass Odoo's security model. It opens a door for data leaks, so IMHO it should be `True` by default.

The behavioural change is true, though. That can be fixed with a migration script.

Comment (pedrobaeza) [reply dms/models/storage.py:91]: As said, this should continue as it was before, with no change. If you want it, then you enable it.

Comment (yajo) [reply dms/models/storage.py:91]: Could you please elaborate why?

Comment (pedrobaeza) [reply dms/models/storage.py:91]: I have already told you. It's this way also in 12.0. And even more: this only makes sense if the storage is attachment. I would accept an onchange if you change the storage type to attachment.

Comment (yajo) [reply dms/models/storage.py:91]: Ah OK, @victoralmau could you please add the onchange?

**Review by pedrobaeza — APPROVED** [review]

**Review by yajo — APPROVED** [review]

**Pull request merged**